### DISCUSSION TITLE
Building a Quantizer Node

### DISCUSSION THREAD

**Discussion by AlasdairMott** [maximecb/noisecraft#75] [Q&A] [answered]
I'm trying to write a cv quantizer node. The basic idea is to map input audio to the nearest note in a scale.

 I've written a working audio node, but I'm having trouble figuring out how to enter a scale. I figured I could enter a list of 1v/octave cv notes as a string like this:
 ```
'Quantize':{
        ins: [
            { name: 'in', default: 0 },
        ],
        outs: ['out'],
        params: [
            { name: 'notes', default: "0,0.25,0.417,0.583,0.833" },
        ],
        state: [],
        description: 'quantize input to nearest value in quantization list',
    }, 
```
But when editing the field the string is limited to a small character count and I get this error:
<img width="438" alt="Screen Shot 2022-04-28 at 8 11 23 am" src="https://user-images.githubusercontent.com/25563790/165639808-166fe620-97ab-4a9f-ad12-0efaf324bb81.png">

A dropdown menu of user defaults would be preferable but I'm not sure how to add that either.

Aside from the user enter considerations, I think a quantizer node might also require a CVToFrequency node and a FrequencyToCV node. I'm currently doing the conversion inside of the quantizer node although I think separate nodes would add more flexibility.

Anyway this is what the quantizer audio node looks like:
```
/**
 * Quantize node
 */
class Quantize extends AudioNode {
    constructor(id, state, sampleRate, send) {
        super(id, state, sampleRate, send);
    }

    frequencyToCV = (freq) => Math.log2(freq / 55.0);

    CVToFrequency = (cv) => Math.pow(2, cv) * 55;

    update(input, notes) {
        if (input < 0) { return 0 }

        //split the string list and convert to doubles
        const noteList = notes.split(',').map(note => parseFloat(note));

        //convert the input to a cv value
        const cv = this.frequencyToCV(input);

        //take the decimal part of the cv value
        const octave = Math.floor(cv);
        const decimal = cv - octave;

        //find the closest note
        const closestNote = noteList.reduce((prev, curr) => {
            return (Math.abs(curr - decimal) < Math.abs(prev - decimal)) ? curr : prev;
        });

        //convert back to a frequency
        return this.CVToFrequency(octave + closestNote);
    }
}
```

**Comment by maximecb** [comment, marked as answer]
Hi Alasdair. Currently for nodes like `MonoSeq`, the scale is represented by 3 state variables ('scaleName', 'scaleRoot', 'numOctaves'). Though the code to render the drop-down menu is specialized for that node type.

The reason you get an error for `notes` is because of the validation code in `validateParams` here:
https://github.com/maximecb/noisecraft/blob/main/public/model.js#L630

It infers the type of the value based on its default. It's not really built to handle a list of values. You'd have to add another case to handle an array as the default value.

**Reply by AlasdairMott** [reply]
Thanks for the advice! I've added a dropdown similar to the mono-sequencer
<img width="279" alt="image" src="https://user-images.githubusercontent.com/25563790/166088027-c8c5e509-abdd-44bb-b658-3377fc07cfc3.png">
